### PR TITLE
Type components in object type syntax (#2359) (restored and extended)

### LINKED ISSUE #760: Support type components in object type syntax
AS currently has no way to specify type components in object *types*

```
ActorScript 0.1 interpreter
> object{public type foo = Int};
{} : {type foo = Int}
> type t = {type foo = Int}
stdin:2.11-2.15: syntax error, unexpected token
```

### PR DESCRIPTION
**Pull request by crusso**
This resurrects @nomeata 's abandoned implementation of type components in the syntax of (object/actor/module) types

That PR only allowed (argument free) type abbreviations, not type constructors, which I'll add here.

Supporting type components in syntactic object types is an independently useful feature that will also fix issue #3449.

In detail (copying from `language-manual.md`):

``` bnf
<typ-field> ::=                               object type fields
  <id> : <typ>                                  immutable value
  var <id> : <typ>                              mutable value
  <id> <typ-params>? <typ1> : <typ2>            function value (short-hand)
  type <id> <typ-params>? = <typ>               type component
```

A type field specifies the name and type of a value field of an object, or the name and definition of a type component of an object. The value field names within a single object type must be distinct and have non-colliding hashes. The type component names within a single object type must also be distinct and have non-colliding hashes. Value fields and type components reside in separate name spaces and thus may have names in common.

`<id> : <typ>` specifies an *immutable* field, named `<id>` of type `<typ>`.

`var <id> : <typ>` specifies a *mutable* field, named `<id>` of type `<typ>`.

`type <id> <typ-params>? = <typ>` specifies a *type* component, with field name `<id>`, abbreviating (parameterized) type `<typ>`.

Unlike type declarations, a type component is not, in itself, recursive (though it may abbreviate an existing recursive type).
In particular, the name `<id>` is not bound in `<typ>` nor in any other fields of the enclosing object type. The name `<id>` only determines the label to use when accessing the definition through a record of this type (using the dot notation).

### CODE REVIEW

File: doc/md/language-manual.md [hunk 1]
```diff
@@ -818,17 +818,24 @@ In all other positions, `( <typ> )` has the same meaning as `<typ>`.
 
 ``` bnf
 <typ-field> ::=                               object type fields
-  <id> : <typ>                                  immutable
-  var <id> : <typ>                              mutable
-  <id> <typ-params>? <typ1> : <typ2>            function (short-hand)
+  <id> : <typ>                                  immutable value
+  var <id> : <typ>                              mutable value
+  <id> <typ-params>? <typ1> : <typ2>            function value (short-hand)
+  type <id> <typ-params> = <typ>                type component
```
Comment (crusso) [inline doc/md/language-manual.md:824]:
```suggestion
  type <id> <typ-params>? = <typ>                type component
```

File: doc/md/language-manual.md [hunk 2]
```diff
@@ -818,17 +818,24 @@ In all other positions, `( <typ> )` has the same meaning as `<typ>`.
...
 ```
 
-A type field specifies the name and type of a field of an object. The field names within a single object type must be distinct and have non-colliding hashes.
+A type field specifies the name and type of a value field of an object, or the name and definition of a type component of an object. The value field names within a single object type must be distinct and have non-colliding hashes. The type component names within a single object type must also be distinct and have non-colliding hashes. Value fields and type components reside in separate name spaces and thus may have names in common.
 
 `<id> : <typ>` specifies an *immutable* field, named `<id>` of type `<typ>`.
 
 `var <id> : <typ>` specifies a *mutable* field, named `<id>` of type `<typ>`.
 
+`type <id> <typ-params> = <typ>` specifies a *type* component, with field name `<id>`, abbreviating (parameterized) type `<typ>`.
```
Comment (crusso) [inline doc/md/language-manual.md:833]:
```suggestion
`type <id> <typ-params>? = <typ>` specifies a *type* component, with field name `<id>`, abbreviating (parameterized) type `<typ>`.
```

File: doc/md/language-manual.md [hunk 3]
```diff
@@ -818,17 +818,24 @@ In all other positions, `( <typ> )` has the same meaning as `<typ>`.
 
 ``` bnf
 <typ-field> ::=                               object type fields
-  <id> : <typ>                                  immutable
-  var <id> : <typ>                              mutable
-  <id> <typ-params>? <typ1> : <typ2>            function (short-hand)
+  <id> : <typ>                                  immutable value
+  var <id> : <typ>                              mutable value
+  <id> <typ-params>? <typ1> : <typ2>            function value (short-hand)
+  type <id> <typ-params>? = <typ>                type component
```
Comment (crusso) [inline doc/md/language-manual.md:824]:
```suggestion
  type <id> <typ-params>? = <typ>               type component
```

File: test/run-drun/issue-3449.mo [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+
+  private type List<A> = ?(A,List<A>);
+
+  private type Bound<A <: {}> = {};
+
+  stable var v : [
+    actor {
+      type T = Int;
+      type U<A> = (A, A);
+      type List<A> = List<A>; // references previous List type, not this List component (type components arent recursive, let alone mutually recursive)
```
Comment (crusso) [inline test/run-drun/issue-3449.mo:14]:
```suggestion
      type List<A> = List<A>; // references previous List type, not this List component (type components aren't recursive, let alone mutually recursive)
```

File: doc/md/language-manual.md [hunk 5]
```diff
@@ -818,17 +818,24 @@ In all other positions, `( <typ> )` has the same meaning as `<typ>`.
...
+A type field specifies the name and type of a value field of an object, or the name and definition of a type component of an object. The value field names within a single object type must be distinct and have non-colliding hashes. The type component names within a single object type must also be distinct and have non-colliding hashes. Value fields and type components reside in separate name spaces and thus may have names in common.
 
 `<id> : <typ>` specifies an *immutable* field, named `<id>` of type `<typ>`.
 
 `var <id> : <typ>` specifies a *mutable* field, named `<id>` of type `<typ>`.
 
+`type <id> <typ-params>? = <typ>` specifies a *type* component, with field name `<id>`, abbreviating (parameterized) type `<typ>`.
+
+Unlike type declarations, a type component is not, in itself, recursive (though it may abbreviate an existing recursive type).
+In particular, the name `<id>` is not bound in `<typ>` nor in any other fields of the enclosing object type. The name `<id>` only determiness the label to use when accessing the definition through a record of this type (using the dot notation).
```
Comment (ggreif) [inline doc/md/language-manual.md:836]:
```suggestion
In particular, the name `<id>` is not bound in `<typ>` nor in any other fields of the enclosing object type. The name `<id>` only determines the label to use when accessing the definition through a record of this type (using the dot notation).
```

File: src/mo_frontend/typing.ml [hunk 6]
```diff
@@ -223,6 +223,23 @@ let system_funcs tfs =
...
+    | T.Abs( _, _) -> true in
+  let typ_params = T.ConSet.filter is_typ_param env.cons in
+  let cs_k = T.cons_kind k in
+  let free_params = T.ConSet.inter typ_params cs_k in
+  if not (T.ConSet.is_empty free_params) then
+    let op, sbs, st = T.strings_of_kind k in
+    error env at "M0137"
+      "type %s%s %s %s references type parameter%s %s from an outer scope"
+      id.it sbs op st
+      (if T.ConSet.cardinal free_params = 1 then "" else "s")
```
Comment (ggreif) [inline src/mo_frontend/typing.ml:240]: this could go into a helper in `T.plural` (are there other occurrences?)

Comment (ggreif) [reply src/mo_frontend/typing.ml:240]: Ha!
``` OCaml
src/mo_frontend/typing.ml:635:      (if T.ConSet.cardinal cs = 1 then "" else "s")
src/mo_frontend/typing.ml:2532:      (if T.ConSet.cardinal free_params = 1 then "" else "s")
```

File: test/fail/bad-type-comp.mo [hunk 7]
```diff
@@ -0,0 +1,3 @@
+ignore ((module {}) : module { type T = Null });
+ignore ((module { public type T = Int}) : module { type T = Null });
```
Comment (ggreif) [inline test/fail/bad-type-comp.mo:2]: What about a self-recursive `type T = T`?

File: test/fail/type-comp.mo [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+  type Upper<A <: {}> = A;
+  type B1 = module { type T<P1 <: { x : Int}> = Upper<P1> }; // accept
+  type B2 = module { type T<P2 <: Null> = Upper<P2> }; // reject
+  type B3 = module { type T<P3> = Upper<P3> }; // reject
+};
+
+do {
+  type List<A> = ?(A,List<A>);
+  type R1 = module { type List<A> = List<A> }; // accept (non-recursive, refers to previous List)
+  type R2 = module { type Bogus<A> = Bogus<A> }; // reject (non-recursive, referes to nothing)
```
Comment (ggreif) [inline test/fail/type-comp.mo:20]:
```suggestion
  type R2 = module { type Bogus<A> = Bogus<A> }; // reject (non-recursive, refers to nothing)
```

File: test/run-drun/issue-3449.mo [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+
+  private type Bound<A <: {}> = {};
+
+  stable var v : [
+    actor {
+      type T = Int;
+      type U<A> = (A, A);
+      type List<A> = List<A>; // references previous List type, not this List component (type components aren't recursive, let alone mutually recursive)
+      type Pair<A,B> = (A,B);
+      type Bound<A<:{}> = Bound<A>; // ditto
```
Comment (ggreif) [inline test/run-drun/issue-3449.mo:16]: suggest spaces
```suggestion
      type Bound<A <: {}> = Bound<A>; // ditto
```

File: src/mo_types/type.ml [hunk 10]
```diff
@@ -704,44 +704,54 @@ let rec span = function
...
-let rec cons' t cs =
+(* Parameter `inTyp` controls whether to count a constructor, `c`,  that only occurs as an argument of
+   `Typ` field, `Typ c`, but not in its own unfolding.
+   Set to false to avoid emitting redundant bindings in stable signature.
+*)
+
+let rec cons' inTyp t cs =
   match t with
   | Var _ ->  cs
   | (Prim _ | Any | Non | Pre ) -> cs
```
Comment (ggreif) [inline src/mo_types/type.ml:715]:
```suggestion
  | Prim _ | Any | Non | Pre -> cs
```

File: src/mo_frontend/typing.ml [hunk 11]
```diff
@@ -2723,15 +2736,24 @@ let check_stab_sig scope sig_ : (T.field list) Diag.result =
...
           let env1 = adjoin env scope in
           check_ids env "object type" "field"
-            (List.map (fun (field : typ_field) -> field.it.id) sfs);
+            (List.filter_map (fun (field : typ_field) ->
+                 match field.it with ValF (id, _, _) -> Some id | _ -> None)
+               sfs);
+          check_ids env "object type" "type field"
+            (List.filter_map (fun (field : typ_field) ->
+                 match field.it with TypF (id, _, _) -> Some id | _ -> None)
+               sfs); (*TODO: reject/assert instead *)
```
Comment (ggreif) [inline src/mo_frontend/typing.ml:2745]: I don't understand this TODO

File: Changelog.md [hunk 12]
```diff
@@ -2,7 +2,14 @@
...
 * motoko (`moc`)
 
-  * halve (default ir-checking) compilation times by optimizing type comparison and hashing (#3463)
+    * halve (default ir-checking) compilation times by optimizing type comparison and hashing (#3463)
+
+    * Add support for type components in object type syntax (#3457, also fixes #3449)
+    ``` motoko
+      type Record = { type T = Nat; x : Nat};
+    ```
+    is now legal. Note the definition of `T` is neither recursive, nor bound in `x : Nat`, but can refer to an existing recursive type.
```
Comment (ggreif) [inline Changelog.md:11]: ...from outer scope

**Review by ggreif — COMMENTED** [review]
Looks good after first reading.

File: test/fail/bad-type-comp.mo (hunk 7 above)
Comment (ggreif) [reply test/fail/bad-type-comp.mo:2]: And maybe a nested type trying to use it.

File: Changelog.md (hunk 12 above)
Comment (crusso) [inline Changelog.md:11]:
```suggestion
    is now legal. Note the definition of `T` is neither recursive, nor bound in `x : Nat`, but can refer to an existing recursive type declared in an outer scope.
```

File: src/mo_types/type.ml (hunk 10 above)
Comment (crusso) [reply src/mo_types/type.ml:715]: Done

File: Changelog.md (hunk 12 above)
Comment (crusso) [reply Changelog.md]: Done

**Comment by crusso** [comment]
@ggreif PTAL

File: test/fail/type-comp.mo [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+};
+
+do {
+  type Outer = module { type Open<A, B> = (A, B) }; // accept (closed type definition)
+};
+
+do {
+  type Outer<A> = module { type Open<B> = (A, B) }; // reject (open type definition)
+};
+
```
Comment (crusso) [inline test/fail/type-comp.mo:30]:
```suggestion
```

**Pull request merged**